### PR TITLE
[ROCm] NGramRepeatBlock, LongformerAttention and DecoderAttention Ops

### PR DESCRIPTION
**Pull request by xinyazhang**
**Description**: Add NGramRepeatBlock, LongformerAttention and DecoderAttention Op to ROCm and/or enable their tests

**Motivation and Context**
- Why is this change required? What problem does it solve?
    * Add and test NGramRepeatBlock, LongformerAttention and DecoderAttention Ops.

### CODE REVIEW

**Comment by ytaous** [comment]
can u pls resolve conflicts first? thx

**Comment by xinyazhang** [comment]
> can u pls resolve conflicts first? thx

Updated

**Comment by ytaous** [comment]
pls fix lint error

File: tools/ci_build/amd_hipify.py [hunk 1]
```diff
@@ -286,10 +276,13 @@
~        "#include <hipcub/backend/rocprim/iterator/discard_output_iterator.hpp>",
~    )
~    s = s.replace("#include <cub/util_allocator.cuh>", "#include <hipcub/util_allocator.hpp>")
     s = s.replace('#include "cub/util_allocator.cuh"', "#include <hipcub/util_allocator.hpp>")
```
Comment (ytaous) [inline tools/ci_build/amd_hipify.py:279]: maybe install "black" and see why it's complaining

Comment (xinyazhang) [reply tools/ci_build/amd_hipify.py:279]: I didn't get it. What is the "black".
Btw the indents actually align well and I'm not sure GitHub can't display correctly.

Comment (ytaous) [reply tools/ci_build/amd_hipify.py:279]: black is the python format tool used in the pipeline - pip install black
https://pypi.org/project/black/

Comment (xinyazhang) [reply tools/ci_build/amd_hipify.py:279]: Thank you!
It did change something, although I'm not sure what it changed

**Comment by xinyazhang** [comment]
I think the C++ lint error is a false alert. Those lines do not exceed the 120-character limit
The Python is the conventional `s = s.replace(...)` practice.

File: onnxruntime/contrib_ops/rocm/bert/attention_impl.cu [hunk 2]
```diff
@@ -170,6 +170,172 @@ bool QkvToContext(
...
+    }
+  }
+
+  // scratch3 is BxNxSxH, transpose to output SxBxNxH
+  return LaunchTransCtx(stream, sequence_length, batch_size, head_size,
+                        num_heads, max_threads_per_block, true, scratch3, output);
+}
+
+
 bool LaunchAttentionKernel(
```
Comment (ytaous) [inline onnxruntime/contrib_ops/rocm/bert/attention_impl.cu:339]: pls move this up so it's easier to compare with cuda's

File: onnxruntime/test/common/rocm_op_test_utils.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
```
Comment (ytaous) [inline onnxruntime/test/common/rocm_op_test_utils.h:4]: maybe don't add this file - 
leverage USE_ROCM or something like:
bool enable_rocm = (nullptr != DefaultRocmExecutionProvider().get()) 

if in future checking version is required, we can do:
#if defined(USE_CUDA) || (defined(USE_ROCM)  && ROCM_VERSION >= 40300)

File: onnxruntime/test/contrib_ops/decoder_attention_op_test.cc [hunk 4]
```diff
@@ -36,9 +37,10 @@ static void RunAttentionTest(
 ) {
   int min_cuda_architecture = use_float16 ? 530 : 0;
   bool enable_cuda = HasCudaEnvironment(min_cuda_architecture);
+  bool enable_rocm = HasRocmEnvironment(min_cuda_architecture);
```
Comment (ytaous) [inline onnxruntime/test/contrib_ops/decoder_attention_op_test.cc:40]: change to use USE_ROCM or (nullptr != DefaultRocmExecutionProvider().get())

**Comment by ytaous** [comment]
Same question here, what's the use case for enabling these OPs?

**Comment by ytaous** [comment]
Hi any update on this? No hurry, we can try to close the others before this one, thx.

**Comment by xinyazhang** [comment]
> Hi any update on this? No hurry, we can try to close the others before this one, thx.

Working on PR #11968 and #11972 now. Will go back to this one after pushing necessary changes.

**Comment by xinyazhang** [comment]
``` bash
$ (cd ./branch_build/$(git branch --show-current)/RelWithDebInfo/; ./onnxruntime_test_all --gtest_filter='NGramRepeatBlockTest.*:LongformerAttentionTest.*:DecoderAttentionTest.*:')
...
[----------] Global test environment tear-down
[==========] 13 tests from 3 test suites ran. (4578 ms total)
[  PASSED  ] 13 tests.
```
GPU Utilization confirmed with `AMD_LOG_LEVEL=3`.

Full log files:
[pool_v2.log](https://github.com/microsoft/onnxruntime/files/9247174/pool_v2.log)
[pool_v2_gpu.log](https://github.com/microsoft/onnxruntime/files/9247175/pool_v2_gpu.log)

**Comment by ytaous** [comment]
fyi - https://github.com/microsoft/onnxruntime/pull/12435

**Comment by ytaous** [comment]
> fyi - #12435

it's merged, can u please update it as needed? thx

**Comment by ytaous** [comment]
heads up - https://github.com/microsoft/onnxruntime/pull/12448

**Comment by ytaous** [comment]
@iK1D - pls also take a look, thx

**Review by ytaous — APPROVED** [review]

**Comment by ytaous** [comment]
Thanks for your contribution.
FYI - @zhangyaobit

File: tools/ci_build/amd_hipify.py [hunk 5]
```diff
@@ -205,7 +193,9 @@ def hipify(src_file_path, dst_file_path):
     if not os.path.exists(dir_name):
         os.makedirs(dir_name, exist_ok=True)
     # Run hipify-perl first, capture output
-    s = subprocess.run([get_hipify_path(), src_file_path], stdout=subprocess.PIPE, universal_newlines=True).stdout
+    s = subprocess.run(
+        [get_hipify_path(), "-roc", src_file_path], stdout=subprocess.PIPE, universal_newlines=True, check=False
```
Comment (zhangyaobit) [inline tools/ci_build/amd_hipify.py:197]: @xinyazhang or @ytaous,  I wonder what is the reason to add -roc here? Thanks!

Comment (ytaous) [reply tools/ci_build/amd_hipify.py:197]: I found this: https://github.com/ROCm-Developer-Tools/HIPIFY/issues/504

**Pull request merged**